tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -11912,6 +11981,1374 @@ VERIFY_SUCCEEDED(DoArraysMatch<T>(OutputVector, ExpectedVector,
                                   TestConfig.Tolerance));
 }
 
+// Runs a set of tests for the Cooperative Vector Mul and MulAdd operations.
+// The device will be queried for supported configurations and then each
+// supported configuration will be tested against multiple matrix and vector
+// sizes. To help reproduce individual test failures, the test will log the
+// configuration it is running and the results of each test. The following
+// filters can be used to limit test execution to a specific set of
+// configurations:
+//
+// - CoopVecMatrixInterp: SINT8, FLOAT16, FLOAT_E4M3, ...
+// - CoopVecMatrixLayout: ROW_MAJOR, COLUMN_MAJOR, MUL_OPTIMAL,
+// OUTER_PRODUCT_OPTIMAL
+// - CoopVecBiasInterp: SINT32, FLOAT16, FLOAT_E4M3, ...
+// - CoopVecInputInterp: SINT8, FLOAT16, FLOAT_E4M3, ...
+// - CoopVecInputType: SINT8, UINT8, SINT16, UINT16, SINT32, UINT32, FLOAT16,
+// FLOAT32, ...
+// - CoopVecOutputType: SINT32, UINT32, FLOAT16, FLOAT32, ...
+//
+// Filter example:
+//   TE.exe ... -p:CoopVecMatrixInterp=FLOAT16
+//   -p:CoopVecMatrixLayout=MUL_OPTIMAL
+//
+// The current implementation will always write the final output data as float.
+void ExecutionTest::runCoopVecMulTest() {
+#if !HAVE_COOPVEC_API
+  WEX::Logging::Log::Comment(
+      "Cooperative vector API not supported in build configuration. Skipping.");
+  WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+  return;
+#else
+  // Create device and verify coopvec support
+  CComPtr<ID3D12Device> D3DDevice;
+  if (!CreateDevice(&D3DDevice, D3D_SHADER_MODEL_6_9)) {
+    return;
+  }
+  if (!DoesDeviceSupportCooperativeVector(D3DDevice)) {
+    WEX::Logging::Log::Comment(
+        "Device does not support cooperative vector. Skipping.");
+    WEX::Logging::Log::Result(WEX::Logging::TestResults::Skipped);
+    return;
+  }
+
+  // Query coopvec feature data. First call gets the size of the arrays. The
+  // second call populates the arrays using memory we allocate.
+  D3D12_FEATURE_DATA_COOPERATIVE_VECTOR DevOptions = {};
+  VERIFY_SUCCEEDED(D3DDevice->CheckFeatureSupport(
+      (D3D12_FEATURE)D3D12_FEATURE_COOPERATIVE_VECTOR, &DevOptions,
+      sizeof(DevOptions)));
+
+  // Allocate memory for the arrays in DevOptions
+  std::vector<D3D12_COOPERATIVE_VECTOR_PROPERTIES_MUL> MulAddProps(
+      DevOptions.MatrixVectorMulAddPropCount);
+  DevOptions.pMatrixVectorMulAddProperties = MulAddProps.data();
+
+  VERIFY_SUCCEEDED(D3DDevice->CheckFeatureSupport(
+      (D3D12_FEATURE)D3D12_FEATURE_COOPERATIVE_VECTOR, &DevOptions,
+      sizeof(DevOptions)));
+
+  // Test each supported data type and matrix layout
+  for (auto MulAddConfig : MulAddProps) {
+    // Filter on preview test support
+    bool PreviewConfig = false;
+    if (MulAddConfig.InputType == D3D12_LINEAR_ALGEBRA_DATATYPE_FLOAT16 &&
```

Comment:
Just a minor comment, no change really necessary:
Wouldn't a nicer way to do this be to have an array of preview configs to compare against?